draft/celery-kubernetes-operator.rst

```diff
@@ -0,0 +1,273 @@
+.. vale off
+
+============================================================
+CEP XXXX: Celery Kubernetes Operator - Architecture Document
+============================================================
+
+:CEP: XXXX
+:Author: Gautam Prajapati
+:Implementation Team: Gautam Prajapati
+:Shepherd: Omer Katz, Asif Saif Uddin
+:Status: Draft
+:Type: Feature
+:Created: 2020-08-30
+:Last-Modified: 2020-08-30
+
+.. contents:: Table of Contents
+   :depth: 3
+   :local:
+
+Abstract
+========
+
+Kubernetes as a container orchestrator is nowadays a popular choice for deploying applications. To run Celery in production on Kubernetes, there are manual steps involved like -
+
+	* Writing deployment spec for workers
+	* Setting up monitoring using `Flower <https://flower.readthedocs.io/en/latest/>`_ 
+	* Setting up Autoscaling
+
+Apart from that, there’s no standard or consistent way to set up multiple clusters, everyone configures their own way which could create problems for infrastructure teams to manage and audit later.
+
+This project attempts to solve(or automate) these issues. It is aiming to bridge the gap between application engineers and infrastructure operators who manually manage the celery clusters.
+
+Moreover, since Celery is written in Python, we plan to use open-source `KOPF <https://github.com/nolar/kopf>`_\ (Kubernetes Operator Pythonic Framework) to write the custom controller implementation. 
```

Comment:
```suggestion
Moreover, since Celery is written in Python, we plan to use open-source `KOPF <https://github.com/nolar/kopf>`_\ (Kubernetes Operator Pythonic Framework) to write the custom controller implementation. 
```
Should we use the Zalando KOPF? 
https://github.com/zalando-incubator/kopf

Reply:
we can collaborate with https://github.com/nolar/kopf as well, though your concerns are valid. we usually try to offer a helping hand to the projects we rely on. I like the Kopf though O:)

Reply:
@nolar wrote a detailed post on the same - https://medium.com/@nolar/kopf-is-forked-cdca40026ea7
He plans to continue the development on `nolar/kopf` in the future.

Reply:
> we usually try to offer a helping hand to the projects we rely on

That was also my thought, yet it's more to maintain and more to learn

Reply:
Thanks for mentioning and for your support! ;-) 

The previous (original) repository was maintained by one single person too (the same person, actually — me), so there is no big difference with the fork. The 95% commits in the last month is what actually was in my WIP repo for the past year, just finished, tested, pushed, and merged.

Regarding the project maintainability: the plans are to start building a community asap, probably even in 2020 (though, most likely, in 2021). But with zero knowledge on how to do that, this looks like a challenge — I'm reading articles on that now; and trying to choose a name which is not used on GitHub yet. Another option is joining one of the existing communities/organizations with established workflows — needs a comparison of pros & cons. This is something on my table to solve after the current backlog of urgent and pending things is done. But these plans are too fuzzy and uncertain to even mention them in the blog post.

This might be beneficial for the project both short-term, as it gets the boost, and long-term, as it does not become abandoned/unmaintained again.

I hope, this clarifies the future of Kopf for your decision making. If you have any questions — feel free to ask.

Reply:
Apache Software Foundation has an incubator for early projects http://incubator.apache.org . I think going this way has an advantage of established "ways of doing things" (20+ years).

CNCF has an idea of "sandbox projects" https://www.cncf.io/sandbox-projects/ but I'm not that familiar with this one but from what see (from KEDA) the projects are more on their own than in ASF.

Reply:
Here is description of incubation processes:
- ASF http://incubator.apache.org/cookbook/#what_are_the_steps_to_becoming_an_apache_software_foundation_top_level_project
- CNCF https://github.com/cncf/toc/blob/master/process/project_proposals.adoc

@nolar let me know if you would like any help (I'm already PMC of Apache Airflow).